File: android/app/src/main/kotlin/com/poracode/app/model/ProjectWorkspaceModels.kt
package com.poracode.app.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
enum class ProjectFileEntryType {
    @SerialName("file")
    File,

    @SerialName("directory")
    Directory,
}

@Serializable
data class ProjectFileEntry(
    val path: String,
    val name: String,
    val type: ProjectFileEntryType,
    val hasChildren: Boolean? = null,
)

@Serializable
data class ProjectSearchConfig(
    val useIgnoreFiles: Boolean,
    val excludePatterns: List<String>,
)

@Serializable
data class ProjectFileSearchResult(
    val entries: List<ProjectFileEntry>,
    val totalIndexed: Long,
)

@Serializable
data class ProjectTreeResult(
    val directoryPath: String,
    val entries: List<ProjectFileEntry>,
)

@Serializable
data class ProjectTreeSearchResult(
    val entries: List<ProjectFileEntry>,
)

@Serializable
enum class ProjectFileReadStatus {
    @SerialName("ready")
    Ready,

    @SerialName("binary")
    Binary,

    @SerialName("too_large")
    TooLarge,

    @SerialName("unsupported")
    Unsupported,
}

@Serializable
enum class ProjectLineEnding {
    @SerialName("lf")
    Lf,

    @SerialName("crlf")
    Crlf,
}

@Serializable
data class ProjectFileReadResult(
    val path: String,
    val status: ProjectFileReadStatus,
    val modifiedAtMs: Double,
    val content: String? = null,
    val contentBase64: String? = null,
    val lineEnding: ProjectLineEnding? = null,
    val hasBom: Boolean? = null,
)

@Serializable
data class ProjectFileWriteResult(
    val modifiedAtMs: Double,
)

@Serializable
enum class GitStatusDetail {
    @SerialName("summary")
    Summary,

    @SerialName("full")
    Full,
}

@Serializable
enum class GitRemotePlatform {
    @SerialName("github")
    GitHub,

    @SerialName("gitlab")
    GitLab,

    @SerialName("bitbucket")
    Bitbucket,

    @SerialName("unknown")
    Unknown,
}

@Serializable
data class GitRemoteInfo(
    val url: String,
    val platform: GitRemotePlatform,
    val owner: String,
    val repo: String,
)

@Serializable
data class GitFileChange(
    val path: String,
    val oldPath: String? = null,
    val status: String,
    val staged: Boolean,
    val insertions: Long,
    val deletions: Long,
)

@Serializable
data class GitStatusResult(
    val detail: GitStatusDetail? = null,
    val isRepo: Boolean,
    val branch: String,
    val headSha: String? = null,
    val tracking: String,
    val hasRemote: Boolean,
    val remoteInfo: GitRemoteInfo? = null,
    val ahead: Long,
    val behind: Long,
    val staged: List<GitFileChange>,
    val unstaged: List<GitFileChange>,
    val totalInsertions: Long,
    val totalDeletions: Long,
    val mergeInProgress: Boolean? = null,
    val mergeMessage: String? = null,
    val conflictFiles: List<GitFileChange>? = null,
)

@Serializable
data class GitDiffResult(
    val diff: String,
)

@Serializable
data class GitDiffBatchResult(
    val staged: Map<String, String>,
    val unstaged: Map<String, String>,
)

@Serializable
data class GitFileContentResult(
    val oldContent: String,
    val newContent: String,
)

@Serializable
data class GitBranchInfo(
    val name: String,
    val current: Boolean,
    val commit: String,
    val isRemote: Boolean,
    val remote: String? = null,
)

@Serializable
data class GitBranchListResult(
    val current: String,
    val branches: List<GitBranchInfo>,
)

@Serializable
data class GitWorktreeInfo(
    val path: String,
    val branch: String,
    val commit: String,
    val isMain: Boolean,
)

@Serializable
data class GitProjectSnapshotResult(
    val status: GitStatusResult?,
    val branches: GitBranchListResult?,
    val worktrees: List<GitWorktreeInfo>?,
    val ghAvailable: Boolean?,
)

data class ProjectWorkspaceTarget(
    val identity: ProjectIdentity,
    val location: ProjectLocation,
)

File: ios/App/App/Models/ProjectWorkspaceModels.swift
import Foundation

struct ProjectWorkspaceSearchConfig: Codable, Equatable, Sendable {
  let useIgnoreFiles: Bool
  let excludePatterns: [String]
}

enum ProjectWorkspaceEntryType: String, Codable, Equatable, Sendable {
  case file
  case directory
}

struct ProjectWorkspaceEntry: Codable, Equatable, Hashable, Identifiable, Sendable {
  let path: String
  let name: String
  let type: ProjectWorkspaceEntryType
  var hasChildren: Bool?

  var id: String { "\(type.rawValue):\(path)" }
}

struct ProjectFileSearchResult: Codable, Equatable, Sendable {
  let entries: [ProjectWorkspaceEntry]
  let totalIndexed: Int
}

struct ProjectTreeResult: Codable, Equatable, Sendable {
  let directoryPath: String
  let entries: [ProjectWorkspaceEntry]
}

struct ProjectTreeSearchResult: Codable, Equatable, Sendable {
  let entries: [ProjectWorkspaceEntry]
}

enum ProjectFileReadStatus: String, Codable, Equatable, Sendable {
  case ready
  case binary
  case tooLarge = "too_large"
  case unsupported
}

enum ProjectFileLineEnding: String, Codable, Equatable, Sendable {
  case lf
  case crlf
}

struct ProjectFileReadResult: Codable, Equatable, Sendable {
  let path: String
  let status: ProjectFileReadStatus
  let modifiedAtMs: Double
  var content: String?
  var contentBase64: String?
  var lineEnding: ProjectFileLineEnding?
  var hasBom: Bool?
}

struct ProjectFileWriteResult: Codable, Equatable, Sendable {
  let modifiedAtMs: Double
}

enum ProjectGitStatusDetail: String, Codable, Equatable, Sendable {
  case summary
  case full
}

enum ProjectGitRemotePlatform: String, Codable, Equatable, Sendable {
  case github
  case gitlab
  case bitbucket
  case unknown
}

struct ProjectGitRemoteInfo: Codable, Equatable, Sendable {
  let url: String
  let platform: ProjectGitRemotePlatform
  let owner: String
  let repo: String
}

struct ProjectGitFileChange: Codable, Equatable, Hashable, Identifiable, Sendable {
  let path: String
  var oldPath: String?
  let status: String
  let staged: Bool
  let insertions: Int
  let deletions: Int

  var id: String { "\(staged ? "staged" : "unstaged"):\(path):\(oldPath ?? "")" }
}

struct ProjectGitStatus: Codable, Equatable, Sendable {
  var detail: ProjectGitStatusDetail?
  let isRepo: Bool
  let branch: String
  var headSha: String?
  let tracking: String
  let hasRemote: Bool
  let remoteInfo: ProjectGitRemoteInfo?
  let ahead: Int
  let behind: Int
  let staged: [ProjectGitFileChange]
  let unstaged: [ProjectGitFileChange]
  let totalInsertions: Int
  let totalDeletions: Int
  var mergeInProgress: Bool?
  var mergeMessage: String?
  var conflictFiles: [ProjectGitFileChange]?
}

struct ProjectGitDiffResult: Codable, Equatable, Sendable {
  let diff: String
}

struct ProjectGitDiffBatchResult: Codable, Equatable, Sendable {
  let staged: [String: String]
  let unstaged: [String: String]
}

struct ProjectGitFileContentResult: Codable, Equatable, Sendable {
  let oldContent: String
  let newContent: String
}

struct ProjectGitBranchInfo: Codable, Equatable, Hashable, Identifiable, Sendable {
  let name: String
  let current: Bool
  let commit: String
  let isRemote: Bool
  var remote: String?

  var id: String { "\(isRemote ? "remote" : "local"):\(name)" }
}

struct ProjectGitBranchList: Codable, Equatable, Sendable {
  let current: String
  let branches: [ProjectGitBranchInfo]
}

struct ProjectGitWorktreeInfo: Codable, Equatable, Hashable, Identifiable, Sendable {
  let path: String
  let branch: String
  let commit: String
  let isMain: Bool

  var id: String { path }
}

struct ProjectGitSnapshot: Codable, Equatable, Sendable {
  let status: ProjectGitStatus?
  let branches: ProjectGitBranchList?
  let worktrees: [ProjectGitWorktreeInfo]?
  let ghAvailable: Bool?
}
